Add TALOS Humanoid Robot from PAL Robotics

saikishor opened the pull request:
This PR adds the requested models as in the issue: https://github.com/deepmind/mujoco_menagerie/issues/21. The models are generated inspiring from the existing models in the repository.

Please feel free to suggest any suggestions to improve the quality of this PR.

Have a great day,

Best Regards,
Sai Kishor Kothakota

Closes: https://github.com/google-deepmind/mujoco_menagerie/issues/21

saran-t commented:
The license is incorrect, the original repo is LGPL.

saikishor commented:
> The license is incorrect, the original repo is LGPL.

@saran-t Tomorrow, I'll talk to my team and update the proper licensing. Sorry about the issue, recently we have decided to go with Apache 2.0 for all over open-source packages. That's the reason I tagged it. However, I will verify it tomorrow and update it on those repositories too.

Thank you for understanding.

saran-t commented:
Great, that works even better for us :) Thanks!

saikishor commented:
Hello @saran-t,

We have updated the licenses on our repo: https://github.com/pal-robotics/talos_robot/blob/kinetic-devel/LICENSE. I've attached the link for you to look over.

Thank you,

Best Regards,
Sai Kishor Kothakota

amirrazmjoo reviewed (changes requested):
The sequence of actuators does not align with that of the joints. Following the head-related joints, the expected order should be arm_left_1_joint through arm_left_7_joint, but presently it appears in a disorderly fashion. This discrepancy is observed in the right and left arms as well as gripper joints across all control modes, impacting the order of values for 'ctrl' in position control mode as well.

yuvaltassa commented:
@amirrazmjoo Thanks so much for helping with the review!

saikishor commented:
Hello @yuvaltassa  @saran-t  @kevinzakka!

We believe this is ready for review now. 

Thank you!

Best Regards,
PAL Robotics Team

kevinzakka reviewed (changes requested):
Thank you for your contribution to Menagerie! Did a first pass review!

saikishor commented on README.md:
```diff
@@ -32,7 +32,7 @@ well right out of the gate.
 |<img src='assets/google_barkour_vb-barkour_vb.png' width=100>|<img src='assets/unitree_go1-go1.png' width=100>|<img src='assets/unitree_go2-go2.png' width=100>|<img src='assets/boston_dynamics_spot-spot_arm.png' width=100>|<img src='assets/agility_cassie-cassie.png' width=100>|
 |<img src='assets/unitree_g1-g1.png' width=100>|<img src='assets/unitree_h1-h1.png' width=100>|<img src='assets/robotis_op3-op3.png' width=100>|<img src='assets/bitcraze_crazyflie_2-cf2.png' width=100>|<img src='assets/skydio_x2-x2.png' width=100>|
-|<img src='assets/flybody-fruitfly.png' width=100>|<img src='assets/realsense_d435i-d435i.png' width=100>||||
+|<img src='assets/flybody-fruitfly.png' width=100>|<img src='assets/realsense_d435i-d435i.png' width=100>|<img src='assets/pal_talos.png' width=100>||||
```
Done!

saikishor commented on README.md:
```diff
@@ -166,6 +166,38 @@ Menagerie, see [CONTRIBUTING](CONTRIBUTING.md).
 
 ## Menagerie Models
 
+| Robot             | Preview       | Grade   |
```
Done!

saikishor commented on pal_talos/README.md:
```diff
@@ -0,0 +1,31 @@
+## TALOS Description (MJCF)
+
+Requires MuJoCo 2.2.2 or later.
```
Done!

saikishor commented on pal_talos/assets/arm/arm_11.stl:
```diff
@@ -0,0 +1 @@
+arm_1.STL
```
We had to do this due to the following issue: https://github.com/deepmind/mujoco/issues/163. We usually reuse the meshes instead of duplicating them, that's the reason for going with the softlinks.

yuvaltassa replied:
I think this is no longer an issue after [this commit](https://github.com/google-deepmind/mujoco/commit/d12fe594d98ee0500fa6bb7513f3faf3d9f7488a), can you please check is this is still an issue?

We are in the process of merging and would love to avoid all these duplicated files to avoid bloat.

Honestly even if the issue is not resolved I prefer to resolve it first. MuJoCo should not be forcing users to needlessly duplicate files 🙂

saikishor replied:
Hello @yuvaltassa,

These are just simlinks and are not a duplication, so it shouldn't be a problem.

Sure, I can check this out and get back to you

kevinzakka commented:
@yuvaltassa Let's get this merged in with the LEAP hand as well.

kevinzakka commented:
@saikishor There's a missing CLA from either you or @PepMS. Can you take care of that please?

saikishor commented:
> @saikishor There's a missing CLA from either you or @PepMS. Can you take care of that please?

@kevinzakka it seems like when we merged a PR from @PepMS, it added data of the GitHub rather than his original details. I could change the details and push by force. Would that work for you?

kevinzakka commented:
@saikishor Totally fine. If you want, you can also close this PR and make a new one and I'll approve it.

saikishor commented:
> @saikishor Totally fine. If you want, you can also close this PR and make a new one and I'll approve it.

Hello!

I think the CLI should now be green. I've already made the appropriate changes to have it green. Please check and let me know. @kevinzakka Thank you so much for all the support

kevinzakka commented:
@saikishor CLA looks good, waiting on Yuval now!

saikishor commented:
> @saikishor CLA looks good, waiting on Yuval now!

Thank you!

saikishor commented:
> @saikishor Is there a way I can push to this branch to fix some formatting / improve the model so that the import in Google is easier?

Sure! I can give you access to our repositories.

saikishor commented:
@kevinzakka you should've received an invitation to our repository. Please let me know, if that's not the case. 

Thank you!

kevinzakka commented:
@saikishor I made some changes. The velocity version looks pretty untuned, we can take care of it after this gets merged in to reduce headaches. Did not touch the symlinks though.

saikishor commented:
> @saikishor I made some changes. The velocity version looks pretty untuned, we can take care of it after this gets merged in to reduce headaches. Did not touch the symlinks though.

Thank you. I'll try to get them addressed tomorrow morning. :)

kevinzakka commented:
@saikishor Another comment for tomorrow as well, looks like there are a bunch of assets that aren't being used in the model. If we can remove those, it would help reduce the size of the model (and the repo).

saikishor commented:
Thanks for the heads-up!

kevinzakka commented:
Hi @yuvaltassa and @saikishor, I cleaned up all the duplicated assets :)

saikishor commented:
> Hi @yuvaltassa and @saikishor, I cleaned up all the duplicated assets :)

Awesome. Thank you so much! I suppose all the symlinks are also fixed with this right?

kevinzakka commented:
@saikishor Yeah there was no need for symlinks, just assign a different name to the geom.

kevinzakka commented:
@saikishor Do you want to make a brand new PR with just the latest commit so that we don't pollute the history with all the assets? I can close this PR and have you make a new one that way you retain authorship.

saikishor commented:
> @saikishor Do you want to make a brand new PR with just the latest commit so that we don't pollute the history with all the assets? I can close this PR and have you make a new one that way you retain authorship.

@kevinzakka Sure. Tomorrow I can squash all the commits into a single one and then open a new separate PR. I really appreciate all the help. 😊

kevinzakka commented:
@saikishor A note for a future CL, we should add an IMU site and related sensors.

saikishor commented:
> @saikishor A note for a future CL, we should add an IMU site and related sensors.

Sure. Thank you.

kevinzakka commented:
@saikishor Oh and forgot to mention, I'm done pushing PRs to your branch now!

saikishor commented:
> @saikishor Oh and forgot to mention, I'm done pushing PRs to your branch now!

Thank you. Just to wrap it. I would need to do the following right?

* Squash the commits
* Recheck the meshes part
* Add IMU and other sensor sites

Any other changes?

kevinzakka commented:
@saikishor No need to to the IMU now, we can do that in a future followup. I don't want you to do too much more work for this first merge! I think it's simplest to just create a new branch, copy over the files from the last commit I made and make a new PR using that branch. And yeah, maybe you can double check the final result to make sure the robot behaves / looks as expected, I may have introduced a bug (I hope not).

saikishor commented:
> @saikishor No need to to the IMU now, we can do that in a future followup. I don't want you to do too much more work for this first merge! I think it's simplest to just create a new branch, copy over the files from the last commit I made and make a new PR using that branch. And yeah, maybe you can double check the final result to make sure the robot behaves / looks as expected, I may have introduced a bug (I hope not).

Thank you so much! I'm sure your changes are perfect 😉
